streamer/m3u8_concater.py

```diff
@@ -0,0 +1,957 @@
+      line = media_playlist.readline()
+      while line:
+        # Capture the M3U tag, PlaylistType, and ExtVersion.
+        if line.startswith(MediaPlaylist.HEADER_TAGS):
+          MediaPlaylist.header += line
+        line = media_playlist.readline()
+  
+  @staticmethod
+  def inf_is_vid(var_playlists: List['MediaPlaylist']) -> bool:
+    """This is useful to detect whether the stream variant playlists in some
+    master playlist are video playlists or not.  They could be audio playlists
+    as #EXT-X-STREAM-INF because there were no video streams associated
+    with this input.
+    """
+    
+    # NOTE: Ideally, this check should be performed on all the STREAM-INF playlists
+    # for this input, but Shaka-Packager's output guarantees for us that all the
+    # STREAM-INFs will be videos or all will be audios in a master playlist.
+    assert len(var_playlists), ('There MUST be at least one stream variant'
+                                'in a master playlist')
+    # NOTE: Since we intentionally indexed the list, it must be a non-empty list,
+    # i.e. there must be at least one stream variant in each master playlist.
+    return isinstance(var_playlists[0].codec, VideoCodec)
+  
+  @staticmethod
+  def _similar_stream_info(media_playlists: List['MediaPlaylist']
+                                ) -> Dict[str, str]:
+    """A helper method, used to keep only the similar stream_info values and
+    pop out the values that are different between multiple streams.
+    
+    A non-empty list must be passed for the parameter `media_playlist`.
+    """
+    
+    assert len(media_playlists), ('There MUST be at least one media playlist'
+                                  'to collect its stream information')
+    # Get an arbitrary stream info.
+    stream_info = media_playlists[0].stream_info.copy()
+    
+    for key in list(stream_info):
+      for media_playlist in media_playlists:
+        # If a media playlist has a different value for this key, pop it.
+        if media_playlist.stream_info.get(key) != stream_info[key]:
+          stream_info.pop(key)
+          break
+    
+    return stream_info
+  
+  @staticmethod
+  def _get_bandwidth(var_playlists: List['MediaPlaylist'],
+                     durations: List[float]) -> Tuple[str, str]:
+    """A helper method to get the peak and average bandwidth
+    for stream variants.                     .---- dot product operator
+                                             |
+    BANDWIDTH = max((BANDWIDTH)s)            v
+    AVERAGE-BANDWIDTH = (AVERAGE-BANDWIDTH)s . (DURATION)s
+                        /sum((DURATION)s)
+    """
+    
+    band, avg_band = 0, 0.0
+    for var_playlist, duration in zip(var_playlists, durations):
+      band = max(int(var_playlist.stream_info['BANDWIDTH']), band)
+      avg_band += int(var_playlist.stream_info['AVERAGE-BANDWIDTH']) * duration
+    
+    return str(band), str(math.ceil(avg_band/sum(durations)))
```

Comment:
I think a dictionary from the callee and an update from the caller side would be very nice.

For the return types, i try to crowd every thing that will be performed a lot or not pleasant for the eye inside the helper methods so that the concat_xxx methods don't get much bigger or uglier :(